File: Dynamic Array and Ring buffer/lib/dynamic_array.rb
require_relative "static_array"

class DynamicArray
  attr_reader :length

  def initialize
    @store = StaticArray.new(0)
    @length = 0
    @capacity = 8
  end

  # O(1)
  def [](index)
    check_index(index)
    @store[index]
  end

  # O(1)
  def []=(index, value)
    check_index(index)
    @store[index] = value
  end

  # O(1)
  def pop
    check_index
    val = self[@length - 1]
    self[@length - 1] = nil
    @length -= 1
    val
  end

  # O(1) ammortized; O(n) worst case. Variable because of the possible
  # resize.
  def push(val)
    resize! if @length == @capacity
    @length += 1
    self[@length - 1] = val
  end

  # O(n): has to shift over all the elements.
  def shift
    check_index
    val = self[0]
    (1...@length).each do |i|
      self[i-1] = self[i]
    end
    @length -= 1
    val
  end

  # O(n): has to shift over all the elements.
  def unshift(val)
    resize! if @length == @capacity
    @length.downto(0).each do |i|
      @store[i] = @store[i-1]
    end
    @length += 1
    @store[0] = val
  end

  protected
  attr_accessor :capacity, :store
  attr_writer :length

  def check_index(index = 0)
    raise 'index out of bounds' if @length.zero? || !index.between?(0, @length - 1)
  end

  # O(n): has to copy over all the elements to the new store.
  def resize!
    @capacity *= 2
    new_store = StaticArray.new(@capacity)
    (0...@length).each do |i|
      new_store[i] = self[i]
    end
    @store = new_store
  end
end

File: Algorithms class-Coursera/Data Structure/DynamicArray.js

class DynamicArray {
  constructor() {
    this.capacity = 8;
    this.store = new Array(this.capacity);
    this.length = 0;
  }

  get(i) {
    // O(1)
    this.inBoundValidation(i);
    return this.store[i];
  }

  set(i, val) {
    this.inBoundValidation(i);
    if (this.store[i] === undefined) {
      // set new element => addition
      this.ensureLengthAndResize();
      this.length++;
    }
    if (val === undefined) {
      // set element to undefined => deletion
      this.length--;
    }
    this.store[i] = val;
  }

  push(val) {
    // O(1) ammortized
    this.ensureLengthAndResize();
    this.set(this.length, val);
  }

  unshift(val) {
    // O(n) to move all the elements to right
    this.ensureLengthAndResize();
    for (let i=this.length; i>0; i--) {
      this.set(i, this.get(i-1));
    }
    this.set(0, val);
  }

  shift() {
    // O(n) to move all the elements to left
    this.checkEmpty();
    let returnEl = this.get(0);
    for (let i=1; i<=this.length; i++) {
      this.set(i-1, this.get(i));
    }
    return returnEl;
  }

  pop() {
    // O(1)
    this.checkEmpty();
    let returnEl = this.get(this.length - 1);
    this.set(this.length - 1, undefined);
    return returnEl;
  }

  checkEmpty() {
    if (this.length === 0) {
      throw 'Store is empty.';
    }
  }

  ensureLengthAndResize() {
    // O(n) to resize the array
    if (this.length === this.capacity) {
      let newStore = new Array(this.capacity * 2);
      for (let i=0; i<this.capacity; i++) {
        newStore[i] = this.get(i);
      }
      this.capacity *= 2;
      this.store = newStore;
    }
  }

  inBoundValidation(i) {
    if (i < 0 || i >= this.capacity) {
      throw 'Index is not in Bound';
    }
  }
}

module.exports = DynamicArray;
